packages/api/test/e2e/promise-contract.spec.js

```diff
@@ -89,6 +89,21 @@ describe.skip('e2e contracts', () => {
           }
         });
     });
+
+    it('decodes tuple return types', (done) => {
+      expct(address).toBeDefined();
+
+      api.tx.contract
+        .call(address, 12345, 500000, abi.messages.compare_and_get(1, 2))
```

Comment:
Don't see `compare_and_get` in the in-repo ABI?

Reply:
I added this as a dumb function just to get a (u64, u64) return value to test on. Would it be better to have an entirely separate contract instead?

Reply:
Would really like to see a valid ABI, so we are sure -

- the ABI parsing is done correctly
- the e2e testing is done correctly

Adding an additional contract + ABI is fine for these

Reply:
To to expand on this - so basically take a peek at https://github.com/polkadot-js/api/pull/906 - it just adds the wasm for an updated test with the ABI.

So when we finally do `Option<T>` and `Result<T>` (and `Vec<T>` which has the same form), that is in place for tests -

- so the ABI parsing can be updated, i.e https://github.com/polkadot-js/api/blob/master/packages/types/src/ContractAbi.ts - that certainly does not cater for this type of parsing (also not sure of having it in "types", but could convince myself that it either belongs in types or api... well, needs to be _somewhere_)
- Once that is find and it can encode/decode based on the definitions, we can add a specific test that uses the ABI + WASM to test on-chain

I hope that explains my previous comments (slightly) better.